Source/Core/VideoCommon/Assets/CustomResourceManager.h

```diff
@@ -0,0 +1,191 @@
+
+private:
+  // A generic interface to describe an assets' type
+  // and load state
+  struct AssetData
+  {
+    std::unique_ptr<CustomAsset> asset;
+    CustomAssetLibrary::TimeType load_request_time = {};
+    std::set<std::size_t> asset_owners;
+
+    enum class AssetType
+    {
+      TextureData
+    };
+    AssetType type;
+
+    enum class LoadType
+    {
+      PendingReload,
+      LoadFinished,
+      LoadFinalyzed,
+      DependenciesChanged,
+      Unloaded
+    };
+    LoadType load_type = LoadType::PendingReload;
+    bool has_errors = false;
+  };
+
+  // A structure to represent some raw texture data
+  // (this data hasn't hit the GPU yet, used for custom textures)
+  struct InternalTextureDataResource
+  {
+    AssetData* asset_data = nullptr;
+    VideoCommon::GameTextureAsset* asset = nullptr;
+    std::shared_ptr<TextureData> texture_data;
+  };
+
+  void LoadTextureDataAsset(const CustomAssetLibrary::AssetID& asset_id,
+                            std::shared_ptr<VideoCommon::CustomAssetLibrary> library,
+                            InternalTextureDataResource* internal_texture_data);
+
+  template <typename T>
+  T* CreateAsset(const CustomAssetLibrary::AssetID& asset_id, AssetData::AssetType asset_type,
+                 std::shared_ptr<VideoCommon::CustomAssetLibrary> library)
+  {
+    const auto [it, added] =
+        m_asset_id_to_session_id.try_emplace(asset_id, m_session_id_to_asset_data.size());
+
+    if (added)
+    {
+      AssetData asset_data;
+      asset_data.asset = std::make_unique<T>(library, asset_id, it->second);
+      asset_data.type = asset_type;
+      asset_data.has_errors = false;
+      asset_data.load_type = AssetData::LoadType::PendingReload;
+      asset_data.load_request_time = {};
+
+      m_session_id_to_asset_data.insert_or_assign(it->second, std::move(asset_data));
+
+      // Synchronize the priority cache session id
+      m_pending_assets.Prepare();
+      m_loaded_assets.Prepare();
+    }
+    auto& asset_data_from_session = m_session_id_to_asset_data[it->second];
+
+    // Asset got unloaded, rebuild it with the same metadata
+    if (!asset_data_from_session.asset)
+    {
+      asset_data_from_session.asset = std::make_unique<T>(library, asset_id, it->second);
+      asset_data_from_session.has_errors = false;
+      asset_data_from_session.load_type = AssetData::LoadType::PendingReload;
+    }
+
+    return static_cast<T*>(asset_data_from_session.asset.get());
+  }
+
+  // This class defines a cache of least or most recently used assets
+  // It is used to keep track of the least used asset,
+  // to know what assets we can remove when memory is low.
+  // It is also used to find the most recent asset,
+  // to know which assets take priority in terms of loading
+  class LeastRecentlyUsedCache
+  {
+  public:
+    const std::list<CustomAsset*>& Elements() const { return m_asset_cache; }
+
+    void put(u64 asset_session_id, CustomAsset* asset)
```

Comment:
`put` -> `Put` or maybe a better name? Is this marking an asset as recently used. I don't think it's immediately obvious how this interface works.